lxd/ubuntu-pro/watcher.go

```diff
@@ -0,0 +1,252 @@
+				// Interfaces directory created. Add a watcher to it.
+				err := p.watcher.Add(interfacesDir)
+				if err != nil {
+					logger.Error("Error occurred while adding file watcher for ubuntu-pro interface directory", logger.Ctx{"err": err})
+				}
+
+			case event.Name == interfacesDir && event.Has(fsnotify.Remove):
+				// Interfaces directory removed. Whatever the previous value of the auto-attach setting, it is now false.
+				p.guestAttachSetting = GuestAttachSettingOff
+			case event.Name == lxdConfigFile && event.Has(fsnotify.Write):
+				// Settings file was written to. Try to read it. If reading fails, set the value to off.
+				s, err := readLXDConfigFile(lxdConfigFile)
+				if err != nil {
+					logger.Error("Failed to read ubuntu-pro auto-attach configuration", logger.Ctx{"err": err})
+					p.guestAttachSetting = GuestAttachSettingOff
+					continue
+				}
+
+				p.guestAttachSetting = s.GuestAttach
+			case event.Name == lxdConfigFile && event.Has(fsnotify.Remove):
+				// Settings file was removed. Whatever the previous value of the auto-attach setting, it is not false.
+				p.guestAttachSetting = GuestAttachSettingOff
+			}
+
+		case err := <-p.watcher.Errors:
+			logger.Error("Error occurred while watching ubuntu-pro guest attach configuration", logger.Ctx{"err": err})
+		}
+	}
+}
+
+// readLXDConfigFile reads the ubuntu-pro configuration file for LXD auto-attach, returning an error if the configuration
+// is invalid.
+func readLXDConfigFile(lxdConfigFile string) (*SettingsFile, error) {
+	f, err := os.Open(lxdConfigFile)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to open ubuntu-pro configuration file: %w", err)
+	}
+
+	defer f.Close()
+
+	s := SettingsFile{}
+	err = json.NewDecoder(f).Decode(&s)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read ubuntu-pro configuration file: %w", err)
+	}
+
+	err = s.GuestAttach.isValid()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read ubuntu-pro configuration file: %w", err)
+	}
+
+	return &s, nil
+}
+
+// setupWatcher configures a new fsnotify.Watcher for the proWatcher. If a "lxd-config.json" already exists, the current
+// value is read and returned. If the ubuntu-advantage directory does not exist, this function will return an error.
+func setupWatcher(ubuntuAdvantageDir string) (*SettingsFile, *fsnotify.Watcher, error) {
+	// Create a file watcher.
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to create a file watcher: %w", err)
+	}
+
+	// Add /var/lib/ubuntu-advantage to the watcher.
+	err = w.Add(ubuntuAdvantageDir)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed adding ubuntu-advantage directory to file watcher: %w", err)
+	}
+
+	// Check if the interfaces directory exists.
+	interfacesDir := interfacesDir(ubuntuAdvantageDir)
+	_, err = os.Stat(interfacesDir)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, nil, fmt.Errorf("Failed to check if ubuntu-pro interface directory exists: %w", err)
```

Comment:
I suspect we'll need to keep checking if it exists periodically, as the package may be installed which creates the dir after LXD is started

Reply:
The `interfaces` directory is added to the watchlist when it is created. I've bailed out if `/var/lib/ubuntu-advantage` doesn't exist with the assumption being it is not an ubuntu host.

Reply:
I think we need to get clarity around when that dir is created incase its managed by the pro client tool. We may need a different indicator for Ubuntu guest.

Reply:
I'm beginning to think this file watcher approach is incorrect. The pro client has a `config` command (used for setting `lxd_guest_attach`) with a `show` subcommand:
```
root@v1:~# pro config show lxd_guest_attach
lxd_guest_attach off
```

How about the `devlxd` handler checks the instance's `ubuntu_pro.guest_attach` setting. If it is `available` or `on`, we call `pro config show lxd_guest_attach` and parse the output to get the current value. If it is `off` or there is an error we return `off`, otherwise we return the instance setting.

This leaves the pro client free to change it's internal representation of the configuration.

@orndorffgrant is the `pro config show` command considered stable for this purpose?

Reply:
Yes potentially that could be a better way instead. But I don't think we should be calling out to the pro client on every /dev/lxd request for guest attach status as it would be  away for the guests to DoS the host (execing external processes are relatively expensive), additionally if you are running lots of instances and you start/restart them they are all going to query the pro client around the same time.

So we would need some form caching in that scenario, so that we only get the current value at most once per X time period. And this could introduce unexpected behaviour when you enable it on the host and then immediately go into the guest to auto attach and it doesn't work straight away.

Hence the event driven approach was chosen using a config file.

Reply:
Makes sense, thanks.